### PR TITLE
fix(otel): balance otel context attach and detach

### LINKED ISSUE #643: [otel] Balance context attach and detach across plugin lifecycles
## Summary

Both OTel plugins call `opentelemetry.context.attach()` without retaining the returned token or calling `detach()`. This leaves a workflow, invocation, operation, or attempt span active after its intended scope has ended.

The test suites currently reset the OTel context around every test specifically because plugin state otherwise leaks between tests.

## Evidence

- `packages/aws-durable-execution-sdk-python-otel/src/aws_durable_execution_sdk_python_otel/execution_plugin.py`, attaches around lines 218-221, 458, and 506-508
- `packages/aws-durable-execution-sdk-python-otel/src/aws_durable_execution_sdk_python_otel/invocation_plugin.py`, attaches around lines 565 and 620-622
- The autouse fixtures in `tests/test_execution_plugin.py` and `tests/test_invocation_plugin.py` state that the plugins attach spans without detaching and would otherwise leak context between tests

A focused runtime probe completes a non-terminal invocation and observes that `trace.get_current_span()` still returns the prior workflow span after `on_invocation_end`.

## OpenTelemetry requirement

The Context specification defines attach as returning a token used to restore the previous context and states that every attach should have a corresponding detach:

https://opentelemetry.io/docs/specs/otel/context/#attach-context
https://opentelemetry.io/docs/specs/otel/context/#detach-context

## Impact

- Auto-instrumented work after plugin cleanup can be parented to a stale span.
- Ended or abandoned spans can remain current across warm Lambda invocations.
- Log correlation may report stale trace and span identifiers.
- Nested operations overwrite context rather than managing a balanced scope.

## Acceptance criteria

- Every `context.attach()` token owned by the plugin has a corresponding `context.detach()`.
- Invocation cleanup restores the context that was active before plugin invocation start.
- User-function cleanup restores the exact enclosing context without accumulating stale scopes.
- Tests no longer require global context resets to hide plugin lifecycle leaks.
- Tests cover nested child contexts, multiple sequential steps, failures, and warm invocation reuse.

### PR DESCRIPTION
**Pull request by wangyb-A**
## Problem

Both OTel plugins called `opentelemetry.context.attach()` and discarded the returned token, so a workflow, invocation, operation, or attempt span stayed current after its scope ended. `on_user_function_end` then "restored" the enclosing span by attaching *again*, so every operation pushed another unbalanced scope onto the thread's context stack. The test suites reset the global OTel context around every test to hide the leak.

Resolves #643

## Change

Each plugin now records the token for every context it attaches, keyed by the same registry key used for the span, and releases it at the matching lifecycle end:

- `on_user_function_end` **detaches** its own scope instead of attaching the enclosing span again. Detaching restores the exact enclosing context, which is what the re-attach was approximating.
- `ExecutionOtelPlugin`'s invocation-start scope is released during invocation cleanup, so the context active before the invocation is restored — this is the case the issue's runtime probe observed.
- Invocation cleanup also releases anything still open, newest first. That covers a user function that suspends (`SuspendExecution` skips `on_user_function_end`) and an invocation that ends abnormally, so nothing survives into a warm invocation.
- A token recorded on another thread is dropped rather than reset, because a context token can only be reset on the thread that created it and OpenTelemetry logs an error otherwise. In practice the pairs line up: invocation hooks run on the Lambda handler thread, and both user-function hooks for one attempt run on the thread executing user code.

No new classes; the helper trio lives on each plugin next to its existing span-registry helpers.

## Behavior delta

Between top-level operations, `trace.get_current_span()` on the worker thread is now the restored enclosing context rather than the plugin's Invocation span. Emitted log correlation is unchanged: `get_current_span_context()` already falls back to the invocation span in the plugin registry, which is the documented path for top-level handler code. Verified that the set of exported spans is byte-identical before and after the change.

## Tests

- The four autouse `_reset_otel_context` fixtures are replaced with an autouse fixture that asserts each test leaves the OTel context as it found it, so a lifecycle leak fails the suite instead of being masked. That fixture immediately caught six pre-existing mid-lifecycle tests; each now ends its invocation, which exercises the new cleanup path.
- New coverage for both plugins: nested child context unwinding, sequential steps not accumulating scopes, failed steps, suspended user functions, warm invocation reuse with an ambient (ADOT-style) span, and the cross-thread token guard.
- Four `InvocationOtelPlugin` tests that asserted the old stacked behavior were rewritten to assert enclosing-context restoration while `get_current_span_context()` still resolves the invocation span.

## Verification

- `hatch run dev-otel:test` — 135 passed
- `hatch run test:all` — 3233 passed, 2 skipped
- `hatch run types:check` — clean
- `hatch fmt --check` (otel package) — clean
- Throwaway runtime probe driving a real `durable_execution()` (outer step + child context + inner step, two invocations) against both plugins. Before the change it reproduced the issue: context not restored, a plugin span still current after the handler returned. After: context restored, no span left current, no tokens open, zero `Failed to detach context` logs, identical exported spans.

### CODE REVIEW

File: packages/aws-durable-execution-sdk-python-otel/src/aws_durable_execution_sdk_python_otel/invocation_plugin.py [hunk 1]
```diff
@@ -167,19 +172,62 @@ def _attempt_span_key(info: UserFunctionStartInfo | UserFunctionEndInfo) -> str:
...
+    # ------------------------------------------------------------------
+    # Context scope helpers
+    # ------------------------------------------------------------------
+    def _attach_context(self, key: str, new_context: Context) -> None:
+        """Attach a context and remember its token under ``key``."""
+        with self._operation_spans_lock:
+            self._context_tokens[key] = (
+                threading.get_ident(),
+                context.attach(new_context),
+            )
```
[the remark this answers is not in the document]
Comment (wangyb-A) [reply packages/aws-durable-execution-sdk-python-otel/src/aws_durable_execution_sdk_python_otel/invocation_plugin.py:184]: This will be fix in #648

File: packages/aws-durable-execution-sdk-python-otel/src/aws_durable_execution_sdk_python_otel/execution_plugin.py [hunk 2]
```diff
@@ -157,6 +170,46 @@ def _pop_span(self, key: str) -> Span | None:
...
+        line up: invocation hooks run on the Lambda handler thread and
+        user-function hooks run on the thread executing user code.
+        """
+        with self._lock:
+            entry = self._context_tokens.pop(key, None)
+        if entry is None:
+            return
+        thread_ident, token = entry
+        if thread_ident == threading.get_ident():
+            otel_context.detach(token)  # type: ignore[arg-type]
```
[the remark this answers is not in the document]
Comment (wangyb-A) [reply packages/aws-durable-execution-sdk-python-otel/src/aws_durable_execution_sdk_python_otel/execution_plugin.py:199]: tracked in #622  2 plugins should not be used at the same time

[the remark this answers is not in the document]
Comment (zhongkechen) [reply packages/aws-durable-execution-sdk-python-otel/src/aws_durable_execution_sdk_python_otel/execution_plugin.py:199]: This sounds concerning.

**Review by zhongkechen — APPROVED** [review]
